golem-rust/src/transaction.rs

```diff
@@ -0,0 +1,167 @@
+///
+/// Implement this trait and use it within a `transaction` block.
+/// Operations can also be constructed from closures using `operation`.
+pub trait Operation<In, Out, Err>: Clone {
+    fn execute(&self, input: In) -> Result<Out, Err>;
+    fn rollback(&self, input: In);
+}
+
+/// Constructs an `Operation` from two closures: one for executing the operation,
+/// and one for rolling it back
+pub fn operation<In, Out, Err>(
+    execute_fn: impl Fn(In) -> Result<Out, Err> + 'static,
+    rollback_fn: impl Fn(In) + 'static,
+) -> impl Operation<In, Out, Err> {
+    FnOperation {
+        execute_fn: Rc::new(execute_fn),
+        rollback_fn: Rc::new(rollback_fn),
+    }
+}
+
+struct FnOperation<In, Out, Err> {
+    execute_fn: Rc<dyn Fn(In) -> Result<Out, Err>>,
+    rollback_fn: Rc<dyn Fn(In)>,
+}
+
+impl<In, Out, Err> Clone for FnOperation<In, Out, Err> {
+    fn clone(&self) -> Self {
+        Self {
+            execute_fn: self.execute_fn.clone(),
+            rollback_fn: self.rollback_fn.clone(),
+        }
+    }
+}
+
+impl<In, Out, Err> Operation<In, Out, Err> for FnOperation<In, Out, Err> {
+    fn execute(&self, input: In) -> Result<Out, Err> {
+        (self.execute_fn)(input)
+    }
+
+    fn rollback(&self, input: In) {
+        (self.rollback_fn)(input)
+    }
+}
+
+/// Transaction is a sequence of operations that are executed in a way that if any of the
+/// operations or the underlying Golem executor fails, the whole transaction is going to
+/// be retried.
+///
+/// In addition to that, **user level failures** (represented by the `Result::Err` value
+/// of an operation) lead to performing the rollback actions of each already performed operation
+/// in reverse order.
+///
+/// Fatal errors (panic) and external executor failures are currently cannot perform the
+/// rollback actions.
+pub struct Transaction {
+    begin_oplog_index: OplogIndex,
+    rollback_actions: Vec<Box<dyn FnOnce()>>,
+}
+
+impl Transaction {
+    fn new(begin_oplog_index: OplogIndex) -> Self {
+        Self {
+            begin_oplog_index,
+            rollback_actions: Vec::new(),
+        }
+    }
+
+    pub fn add<OpIn: Clone + 'static, OpOut, OpErr>(
+        &mut self,
+        operation: impl Operation<OpIn, OpOut, OpErr> + 'static,
+        input: OpIn,
+    ) -> OpOut {
+        let cloned_op = operation.clone();
+        let cloned_in = input.clone();
+        self.rollback_actions.push(Box::new(move || {
+            cloned_op.rollback(cloned_in);
+        }));
+        match operation.execute(input) {
+            Ok(output) => output,
+            Err(_) => {
+                self.fail();
+                unreachable!()
+            }
+        }
+    }
+
+    pub fn fail(&mut self) {
+        for rollback_action in self.rollback_actions.drain(..).rev() {
+            rollback_action();
+        }
+        set_oplog_index(self.begin_oplog_index);
+    }
+}
+
+pub fn transaction<Out>(f: impl FnOnce(&mut Transaction) -> Out) -> Out {
```

Comment:
We have to carefully think through what we want the user API to be here.

I think the requirements are something like:

1. The transaction is attempted.
2. If there is an error, rollback is attempted. 
3. If the transaction succeeds, then the user gets back the successful result of the transaction.
4. If the transaction fails, due to an error, then the user gets back the failure of the transaction, along with knowledge about whether or not all compensation actions succeeded.

So in my mind, our transaction function looks something like:

```rust
pub fn transaction<Out>(f: impl FnOnce(&mut Transaction) -> Out) -> TransactionResult<Out> {  ... }

enum TransactionResult<Out> {
  Success(Out),
  FailedWithFullRollback,
  FailedWithPartialRollback(<error>),
}
```

Moreover, I am not sure we need `set_oplog_index` if we are using the closure. Because we can pop out of the closure at any time to short circuit and return the rollback error, if any.

Does this make sense?

Reply:
I was thinking (and implemented accordingly) that in case of a failure we want to use Golem's capabilities to recover and retry. This means that if there was an error this function never returns. After the rollback operations are finished, it "crashes" with `set-oplog-index` and retries from the beginning of the transactions. 

I thought that was the primary reason why we added `set-oplog-index` - otherwise would not this be a fully "user-land" feature which is not golem-specific at all?

Reply:
I see, did not think of that case. So why would someone use `golem-rust` to define operations etc. if they don't want anything Golem specific (recovery) to happen instead of just writing their business logic that calls A, B and C and does things in case of failure?

Reply:
Well, we are talking business failure versus infrastructure failure: a checkout API needs to explicitly handle the case that the payment processor fails, whereupon it must revert prior steps in the process (including reservations held on inventory); the payment processor will not just succeed if tried more times; it could have failed because the user entered payment information incorrectly, or because they have reached their account or balance limit, etc. 

I think Golem Rust can contain a lightweight implementation of the saga pattern, which can be used for scenarios like these. But upon further reflection, I think your "transaction never fails" semantics are indeed useful. For example: when you are writing to a database and also to a message broker, the transaction can always succeed eventually and you do want always-retry semantics. But for other cases, I think our value add is the saga pattern itself, which becomes reliable on Golem, but there's nothing Golem specific about that Saga pattern implementation.

Does this make sense or am I overlooking something?

Reply:
Do you think the proposed advanced version with the kv store the etc makes sense as well? As a third level of guarantees?

Reply:
Yes, but I would not do that at this point in time (I think it does not require Golem-specific support, only wasi-keyvalue, and of course the runtime API).

However, I have been thinking about the use of atomic region for a transaction: I do not think this is correct. 

Let's take the checkout example:

1. reserve inventory
2. process the payment
3. store the success
4. create the shipping label & tracking number
5. dispatch the order
6. send the email confirmation

The user wants transactional guarantees around all 6 steps. What does that mean?

1. It means that the execution of these steps must be durable, so that if steps 1 - 2 are executed, and a fault occurs, then after recovery, step 3 will be the next step. Moreover, and unique to Golem, they must be fine-grained, such that if step (1) is actually 20 steps, and a fault occurs somewhere inside those 20 steps, then we'll recover wherever we left off when the fault occurred.
2. It also means that if there is a domain / business error, then compensations are applied automatically in reverse order. An example is that processing the payment fails, because of insufficient balance or overextended credit. If processing the payment falls, then the inventory must be released, which is the compensation for reserve inventory.
3. Overall, and around the whole 6 step process, the user expects: the steps execute completely, successfully, and exactly-once (up to the limit of idempotent APIs); OR the process fails somewhere in the middle, for a business / domain error, in which case, the transaction is rolled back / aborted, and control returns to the user, who can handle the domain / business error in a case-specific way.

In order to achieve these goals, we must not use atomic around the transaction, because that prevents durability around the individual operations that make up the steps.

One can argue atomic should be used around operations. That emulates the Temporal semantic (common to other durable execution platforms). However, I don't think we should do that. If a user wants to do that, they can opt-in. It's better to have fine-grained durability by default, IMO.

Reply:
This makes sense and it also means that if the worker gets restarted during compensation, the compensation process will be continued as expected (because the failure and the already performed compensations are already persisted).

I think it is needed in the other version (which I called `RetryingTransaction` now) where after the compensation the transaction is retried with `set-oplog-index`. This means that for domain errors the transaction is treated as atomic because we drop all the previous oplog entries for the retry. By using the atomic region we say that it should behave the same in case of a non-domain error.